hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/joiner/SpanJoiner.java

```diff
@@ -12,16 +12,17 @@ public interface SpanJoiner {
   SpanJoiner NO_OP_JOINER =
       new SpanJoiner() {
         @Override
-        public <T> Single<Map<T, Span>> joinSpans(
+        public <T> Single<Map<T, Collection<Span>>> joinSpans(
             Collection<T> joinSources, SpanIdGetter<T> spanIdGetter) {
           return Single.just(Collections.emptyMap());
         }
       };
 
-  <T> Single<Map<T, Span>> joinSpans(Collection<T> joinSources, SpanIdGetter<T> spanIdGetter);
+  <T> Single<Map<T, Collection<Span>>> joinSpans(
+      Collection<T> joinSources, SpanIdGetter<T> spanIdGetter);
 
   @FunctionalInterface
   interface SpanIdGetter<T> {
-    Single<String> getSpanId(T source);
+    Single<Collection<String>> getSpanIds(T source);
```

Comment:
So your getter interfaces can return a collection - we want to accept liberally. It's just what we return that should be a list - we need to respect order and the caller should be able to rely on that

Reply:
This was poorly phrased on my part. The getter's return type was correct - it should be liberal (it's what we accept). It's what the _joiner_ returns that should be more conservative (the listmultimap)